### ISSUE TITLE
run viem inside a snap: hangs upon reading from blockchain, when using injected `ethereum` provider

### ISSUE THREAD

**Issue by firnprotocol** [MetaMask/snaps#1750]
I am an able to load [viem](https://viem.sh/) inside of a Snap, but am **unable to read from the blockchain.** metamask version: 11.0.0-flask.0

i have developed a minimal MWE in [this single commit](https://github.com/firnprotocol/template-snap-monorepo/commit/ca2b95cdb24e2570cc059afcbbab043c5e0191f7#diff-b47c5a03603aff4a5a42a6d4ef569a9a62d86f260004c519871d0e26a4276803) to the [`template-snap-monorepo`](https://github.com/firnprotocol/template-snap-monorepo).

**Symptoms.** Weirdly, the exact behavior depends on how I try to read. Viem supports both "normal" reads and "multicalls". These have different symptoms: one hangs; the other fails.

- when I do a normal call, either using [`publicClient.readContract`](https://viem.sh/docs/contract/readContract.html) or [`contract.read.myMethod`](https://viem.sh/docs/contract/readContract.html), or even the raw [`publicClient.call`](https://viem.sh/docs/actions/public/call.html), the call **hangs indefinitely**, until it times out and is killed by metamask. see e.g. [this code](https://github.com/firnprotocol/template-snap-monorepo/blob/ca2b95cdb24e2570cc059afcbbab043c5e0191f7/packages/snap/src/index.ts#L21-L32).
- weirdly, i was able to make progress by explicitly doing a [`publicClient.multicall`](https://viem.sh/docs/contract/multicall.html). see [this code](https://github.com/firnprotocol/template-snap-monorepo/blob/ca2b95cdb24e2570cc059afcbbab043c5e0191f7/packages/snap/src/index.ts#L34-L49). upon running that multicall, i got a large error, which can be seen below:
<details>
<summary>Expand error dump</summary>
[{"status":"failure","error":{"details":"Provided value is not JSON-RPC compatible: Expected the value to satisfy a union of , but received: [object Object].","docsPath":"/docs/contract/readContract","metaMessages":["Raw Call Arguments:"," to: 0xca11bde05977b3631167028862be2a173976ca11\n data: 0x82ad56cb00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000fba3912ca04dd458c843e2ee08967fc04f3579c200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000000418160ddd00000000000000000000000000000000000000000000000000000000000000000000000000000000fba3912ca04dd458c843e2ee08967fc04f3579c200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000002470a08231000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac00000000000000000000000000000000000000000000000000000000"," ","Contract Call:"," address: 0xca11bde05977b3631167028862be2a173976ca11\n function: aggregate3((address target, bool allowFailure, bytes callData)[])\n args: ([{"allowFailure":true,"callData":"0x18160ddd","target":"0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2"},{"allowFailure":true,"callData":"0x70a08231000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac","target":"0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2"}])"],"shortMessage":"An unknown RPC error occurred.","name":"ContractFunctionExecutionError","version":"viem@1.10.12","message":"An unknown RPC error occurred.\n\nRaw Call Arguments:\n to: 0xca11bde05977b3631167028862be2a173976ca11\n data: 0x82ad56cb00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000fba3912ca04dd458c843e2ee08967fc04f3579c200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000000418160ddd00000000000000000000000000000000000000000000000000000000000000000000000000000000fba3912ca04dd458c843e2ee08967fc04f3579c200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000002470a08231000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac00000000000000000000000000000000000000000000000000000000\n \nContract Call:\n address: 0xca11bde05977b3631167028862be2a173976ca11\n function: aggregate3((address target, bool allowFailure, bytes callData)[])\n args: ([{"allowFailure":true,"callData":"0x18160ddd","target":"0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2"},{"allowFailure":true,"callData":"0x70a08231000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac","target":"0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2"}])\n\nDocs: https://viem.sh/docs/contract/readContract.html\nDetails: Provided value is not JSON-RPC compatible: Expected the value to satisfy a union of , but received: [object Object].\nVersion: viem@1.10.12","cause":{"details":"Provided value is not JSON-RPC compatible: Expected the value to satisfy a union of , but received: [object Object].","metaMessages":["Raw Call Arguments:"," to: 0xca11bde05977b3631167028862be2a173976ca11\n data: 0x82ad56cb00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000fba3912ca04dd458c843e2ee08967fc04f3579c200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000000418160ddd00000000000000000000000000000000000000000000000000000000000000000000000000000000fba3912ca04dd458c843e2ee08967fc04f3579c200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000002470a08231000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac00000000000000000000000000000000000000000000000000000000"],"shortMessage":"An unknown RPC error occurred.","name":"CallExecutionError","version":"viem@1.10.12","message":"An unknown RPC error occurred.\n\nRaw Call Arguments:\n to: 0xca11bde05977b3631167028862be2a173976ca11\n data: 0x82ad56cb00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000fba3912ca04dd458c843e2ee08967fc04f3579c200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000000418160ddd00000000000000000000000000000000000000000000000000000000000000000000000000000000fba3912ca04dd458c843e2ee08967fc04f3579c200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000002470a08231000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac00000000000000000000000000000000000000000000000000000000\n\nDetails: Provided value is not JSON-RPC compatible: Expected the value to satisfy a union of , but received: [object Object].\nVersion: viem@1.10.12","cause":{"details":"Provided value is not JSON-RPC compatible: Expected the value to satisfy a union of , but received: [object Object].","shortMessage":"An unknown RPC error occurred.","name":"UnknownRpcError","version":"viem@1.10.12","message":"An unknown RPC error occurred.\n\nDetails: Provided value is not JSON-RPC compatible: Expected the value to satisfy a union of , but received: [object Object].\nVersion: viem@1.10.12","cause":{"code":"ERR_ASSERTION"},"code":-1}},"abi":[{"inputs":[{"components":[{"name":"target","type":"address"},{"name":"allowFailure","type":"bool"},{"name":"callData","type":"bytes"}],"name":"calls","type":"tuple[]"}],"name":"aggregate3","outputs":[{"components":[{"name":"success","type":"bool"},{"name":"returnData","type":"bytes"}],"name":"returnData","type":"tuple[]"}],"stateMutability":"view","type":"function"}],"args":[[{"allowFailure":true,"callData":"0x18160ddd","target":"0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2"},{"allowFailure":true,"callData":"0x70a08231000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac","target":"0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2"}]],"contractAddress":"0xca11bde05977b3631167028862be2a173976ca11","functionName":"aggregate3"}},{"status":"failure","error":{"details":"Provided value is not JSON-RPC compatible: Expected the value to satisfy a union of , but received: [object Object].","docsPath":"/docs/contract/readContract","metaMessages":["Raw Call Arguments:"," to: 0xca11bde05977b3631167028862be2a173976ca11\n data: 0x82ad56cb00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000fba3912ca04dd458c843e2ee08967fc04f3579c200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000000418160ddd00000000000000000000000000000000000000000000000000000000000000000000000000000000fba3912ca04dd458c843e2ee08967fc04f3579c200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000002470a08231000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac00000000000000000000000000000000000000000000000000000000"," ","Contract Call:"," address: 0xca11bde05977b3631167028862be2a173976ca11\n function: aggregate3((address target, bool allowFailure, bytes callData)[])\n args: ([{"allowFailure":true,"callData":"0x18160ddd","target":"0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2"},{"allowFailure":true,"callData":"0x70a08231000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac","target":"0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2"}])"],"shortMessage":"An unknown RPC error occurred.","name":"ContractFunctionExecutionError","version":"viem@1.10.12","message":"An unknown RPC error occurred.\n\nRaw Call Arguments:\n to: 0xca11bde05977b3631167028862be2a173976ca11\n data: 0x82ad56cb00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000fba3912ca04dd458c843e2ee08967fc04f3579c200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000000418160ddd00000000000000000000000000000000000000000000000000000000000000000000000000000000fba3912ca04dd458c843e2ee08967fc04f3579c200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000002470a08231000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac00000000000000000000000000000000000000000000000000000000\n \nContract Call:\n address: 0xca11bde05977b3631167028862be2a173976ca11\n function: aggregate3((address target, bool allowFailure, bytes callData)[])\n args: ([{"allowFailure":true,"callData":"0x18160ddd","target":"0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2"},{"allowFailure":true,"callData":"0x70a08231000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac","target":"0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2"}])\n\nDocs: https://viem.sh/docs/contract/readContract.html\nDetails: Provided value is not JSON-RPC compatible: Expected the value to satisfy a union of , but received: [object Object].\nVersion: viem@1.10.12","cause":{"details":"Provided value is not JSON-RPC compatible: Expected the value to satisfy a union of , but received: [object Object].","metaMessages":["Raw Call Arguments:"," to: 0xca11bde05977b3631167028862be2a173976ca11\n data: 0x82ad56cb00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000fba3912ca04dd458c843e2ee08967fc04f3579c200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000000418160ddd00000000000000000000000000000000000000000000000000000000000000000000000000000000fba3912ca04dd458c843e2ee08967fc04f3579c200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000002470a08231000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac00000000000000000000000000000000000000000000000000000000"],"shortMessage":"An unknown RPC error occurred.","name":"CallExecutionError","version":"viem@1.10.12","message":"An unknown RPC error occurred.\n\nRaw Call Arguments:\n to: 0xca11bde05977b3631167028862be2a173976ca11\n data: 0x82ad56cb00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000fba3912ca04dd458c843e2ee08967fc04f3579c200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000000418160ddd00000000000000000000000000000000000000000000000000000000000000000000000000000000fba3912ca04dd458c843e2ee08967fc04f3579c200000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000002470a08231000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac00000000000000000000000000000000000000000000000000000000\n\nDetails: Provided value is not JSON-RPC compatible: Expected the value to satisfy a union of , but received: [object Object].\nVersion: viem@1.10.12","cause":{"details":"Provided value is not JSON-RPC compatible: Expected the value to satisfy a union of , but received: [object Object].","shortMessage":"An unknown RPC error occurred.","name":"UnknownRpcError","version":"viem@1.10.12","message":"An unknown RPC error occurred.\n\nDetails: Provided value is not JSON-RPC compatible: Expected the value to satisfy a union of , but received: [object Object].\nVersion: viem@1.10.12","cause":{"code":"ERR_ASSERTION"},"code":-1}},"abi":[{"inputs":[{"components":[{"name":"target","type":"address"},{"name":"allowFailure","type":"bool"},{"name":"callData","type":"bytes"}],"name":"calls","type":"tuple[]"}],"name":"aggregate3","outputs":[{"components":[{"name":"success","type":"bool"},{"name":"returnData","type":"bytes"}],"name":"returnData","type":"tuple[]"}],"stateMutability":"view","type":"function"}],"args":[[{"allowFailure":true,"callData":"0x18160ddd","target":"0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2"},{"allowFailure":true,"callData":"0x70a08231000000000000000000000000a5cc3c03994db5b0d9a5eedd10cabab0813678ac","target":"0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2"}]],"contractAddress":"0xca11bde05977b3631167028862be2a173976ca11","functionName":"aggregate3"}}]
</details>

i have no idea whether the two errors (i.e., the hanging, vs. the error dumped above) are related, or completely different bugs.

the contracts i'm reading, and the arguments, are taken straight from the [viem docs](https://viem.sh/docs/contract/readContract.html), and are not at fault.

### Steps To Reproduce

clone [my MWE](https://github.com/firnprotocol/template-snap-monorepo). run `yarn install && yarn start`. navigate to http://localhost:8000/. install Flask. connect to the snap in the top right; then click "send message".

please let me know if i can provide any further info.

**Comment by ziad-saab** [comment]
I tried to extract your code outside of the Snaps platform and run it on codesandbox: it fails with an error. Before we look further into errors on the Snaps side, may I ask you to get this version of the code to work as expected:

https://codesandbox.io/s/dreamy-rui-mqpdmx?file=/src/index.mjs

**Comment by firnprotocol** [comment]
hi, @ziad-saab, thanks for your response. i was able to get your sandbox working by changing the `transport`: from `custom(ethereum)` --> `http()`. (and by downcasting `data` to a `Number` before logging it.) of course, `ethereum` is a snap-specific global variable, which we shouldn't expect to exist in the context of that code sandbox; that explains why the sandbox didn't work.

btw, this actually gave me the idea of doing the same thing inside my own snap—i.e., using the http provider instead. interestingly, this gives a temporary workaround. of course, we're now achieving something different from what we wanted: `http()` exploits the `network-access` endowment, and goes through a Viem-specified RPC; by contrast, `custom(ethereum)` uses the `ethereum-provider` endowment and goes through MetaMasks's (injected) provider.

so this is not a suitable workaround for long-term, though it is interesting.

**Comment by mario-christopher** [comment]
Hi @firnprotocol 
I tried out your code, but tried another path instead.
I ignored interacting with the Contract and instead used `Viem` to read info from the Blockchain.
This worked perfectly and I got back information.
This is the code I used inside the Snap
```
   case 'hello':
      const publicClient = createPublicClient({
        chain: mainnet,
        transport: custom((window as any).ethereum),
      });

      const blockNumber = await publicClient.getBlockNumber()
      console.log('DATA', blockNumber); 

    /// output in console
    /// [Snap: local:http://localhost:8080] DATA 18165190n
```

So, this confirms that `Viem` does indeed work from within a Snap. 👍 
The error you see might be related to the Contract in question or how you call a Contract method using Viem.
Could you please check code around interacting with Contracts from Viem.

**Comment by mario-christopher** [comment]
Hi @firnprotocol 
Looks like there's been an incorrect usage of `transport`.
From the Viem docn, the correct way to use it is `transport: http(<url>)`
https://viem.sh/docs/clients/transports/http.html#usage

So the code would work correctly like this:
```
      const publicClient = createPublicClient({
        chain: mainnet, // ???
        transport: http(),
      });

      const blockNumber = await publicClient.getBlockNumber()
      console.log('BLOCKNUMBER', blockNumber);

      const contract = getContract({
        address: '0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2',
        abi: wagmiAbi,
        publicClient,
      });

      const totalSupply = await publicClient.readContract({
        ...contract,
        functionName: 'totalSupply',
      });
      console.log('TOTALSUPPLY', totalSupply);
```

As you too experimented, it works with `http()`. This is not a workaround, but the intended way of usage.
We cannot replace `http()` with `window.ethereum`.
If no `url` parameter is provided to `http(<url>)`, then the transport will fall back to a public RPC URL on the chain. It is highly recommended to provide an authenticated RPC URL to prevent rate-limiting.

Let me know if this satisfies your use-case.
Thanks.

**Comment by firnprotocol** [comment]
@mario-christopher i'm sorry, but you have completely misunderstood my issue.

i AM ALREADY able to call `getBlockNumber` in viem using the `custom(ethereum)` transport. what doesn't work is reading from a contract.

you "solved" my problem by simply changing my code. can you please confirm that you're able to reproduce **my original issue**?

> As you too experimented, it works with `http()`. This is not a workaround, but the intended way of usage.

i'm sorry, but this is nonsense. http isn't "intended"; it's an alternative which goes through a viem-chosen RPC, as opposed to the injected RPC.

if this is "intended", then, please tell me, what is the point of the `ethereum-provider` endowment?!? why include this endowment at all? why include the global `ethereum` variable?

> The error you see might be related to the Contract in question or how you call a Contract method using Viem.

except it's not—please see the code sandbox [here](https://codesandbox.io/s/beautiful-jasper-cxnnwh), where the same contract reads work.

**Comment by firnprotocol** [comment]
btw, as additional data, i can confirm that the hang DOES happen when i use `http(https://eth-mainnet.g.alchemy.com/v2/00000000000myapikey0000000000000)`, but NOT when i use the unspecified `http()`. so something complicated is going on.

please feel free to reproduce

**Comment by Montoya** [comment]
That could be it. The endpoint has to accept `null` origin. Maybe that's why it doesn't work.

**Comment by FrederikBolding** [comment]
> Maybe that's why it doesn't work.
That doesn't explain it hanging when using the ethereum provider though.

@firnprotocol Does `viem` work with `window.ethereum` when injected by MetaMask?

**Comment by firnprotocol** [comment]
hi @FrederikBolding, thanks for your response.

i can confirm that the relevant code _does work_ in the browser, with the `transport: custom(window.ethereum)`.

**Comment by FrederikBolding** [comment]
@firnprotocol The hanging is definitely unintentional. It seems that it occurs because an error is being thrown by `viem` that is non-serializable, I am fixing that here https://github.com/MetaMask/snaps/pull/1772

The error being thrown however is because we currently disallow passing of non JSON-values to `ethereum.request` and `viem` does pass `undefined` for certain parameters..

**Comment by firnprotocol** [comment]
hi @FrederikBolding, excellent, appreciate you confirming the issue.

what will be the ultimate outcome—an explicit (non-hang) failure, or a success?

btw, would it be helpful to raise this on the viem end? they are pretty receptive to suggestions.

**Comment by Montoya** [comment]
It will be an explicit non-hang failure

And yes it would be helpful to raise this to viem. `undefined` is not a JSON value so it should not be used here, although some other clients support it. You can use a hack workaround to modify the value but it's an ugly solution.

**Comment by firnprotocol** [comment]
cc. @jxom—sorry to spam-tag you, but you might find the above discussion worth perusing..

**Comment by jxom** [comment]
Interesting. This is the first EIP-1193 Provider I have encountered that can’t handle `undefined` as a value. EIP-1193 is a **JavaScript** spec, so shouldn’t it handle any type of JavaScript value?

**Comment by FrederikBolding** [comment]
@firnprotocol This should now be fixed in the latest commit of the extension `develop` branch. Will be going out in a future update to both Flask and stable!

**Comment by firnprotocol** [comment]
i see, great. so it will actually work with the injected provider?

**Comment by FrederikBolding** [comment]
It should, yes!

**Comment by firnprotocol** [comment]
k, cool. in the mean time, i ask that you please update this thread when it gets released into public flask. thanks in advance.

(i need this to work for my users, not just myself 😄)

**Comment by FrederikBolding** [comment]
@firnprotocol Apologies, I forgot to update this thread, but this fix went out in Flask `11.3.0`. Let me know if you have any more problems!

**Comment by firnprotocol** [comment]
k, thanks for letting me know.
